homeassistant/components/switch/rainbird.py

```diff
@@ -0,0 +1,224 @@
+            vol.Required(CONF_ZONE): cv.string,
+            vol.Required(CONF_TRIGGER_TIME): cv.string,
+            vol.Optional(CONF_SCAN_INTERVAL): cv.string,
+        },
+    }),
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up Rain Bird switches over a Rain Bird controller."""
+    server = config.get(CONF_HOST)
+    password = config.get(CONF_PASSWORD)
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
+
+    from pyrainbird import RainbirdController
+
+    controller = RainbirdController(_LOGGER)
+    controller.setConfig(server, password)
+    _LOGGER.info("Rain Bird Controller set to " + str(server))
+
+    rbdevice = RainbirdDevice(hass, controller)
+    initialstatus = rbdevice.update()
+    if initialstatus == -1:
+        _LOGGER.error("Error getting state. Possible configuration issues")
+        raise PlatformNotReady
+    else:
+        _LOGGER.info("Initialized Rain Bird Controller")
+
+    entities = []
+    entities.append(rbdevice)
+    component.add_entities(entities)
+
+    devices = []
+    for dev_id, switch in config.get(CONF_SWITCHES).items():
+        devices.append(RainBirdSwitch(rbdevice, switch, dev_id))
+    add_devices(devices)
+    return True
+
+
+class RainBirdSwitch(SwitchDevice):
+    """Representation of a Rain Bird switch."""
+
+    def __init__(self, rb, dev, dev_id):
+        """Initialize a Rain Bird Switch Device."""
+        self._rainbird = rb
+        self._devid = dev_id
+        self._zone = int(dev.get(CONF_ZONE))
+        self._name = dev.get(CONF_FRIENDLY_NAME, "Sprinker %s" % self._zone)
+        self._state = self.get_device_status()
+        self._duration = dev.get(CONF_TRIGGER_TIME)
+        self._attributes = {
+            "duration": self._duration,
+            "zone": self._zone
+        }
+
+    @property
+    def device_state_attributes(self):
+        """Return state attributes."""
+        return self._attributes
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return True
+
+    @property
+    def name(self):
+        """Get the name of the switch."""
+        return self._name
+
+    def turn_on(self, **kwargs):
+        """Turn the switch on."""
+        self._state = True
+        self._rainbird.start_irrigation(self._zone, self._duration)
+        self._rainbird.setstate(self._zone)
+
+    def turn_off(self, **kwargs):
+        """Turn the switch off."""
+        self._state = False
+        self._rainbird.stop_irrigation()
+        self._rainbird.setstate(0)
+
+    def get_device_status(self):
+        """Get the status of the switch from Rain Bird Controller."""
+        return self._rainbird.state == self._zone
+
+    def update(self):
+        """Update switch status."""
+        self._state = self.get_device_status()
+
+    @property
+    def is_on(self):
+        """Return true if switch is on."""
+        return self._state
+
+
+class RainbirdDevice(Entity):
```

Comment:
Make no entity with this. Move that code into pyrainbird and make a better API interface for home-assistant and make simple switches. You can also store the API Object into `hass.data`. I don't know the rainbird but maybe you can read all available switch from interface and add this into hass without you need do that manual. After that you can use `customize` for change the friendly name.

Reply:
Move it again ;) It took me way to long already to analyse the Rain Bird data and get an API of some sorts. I've already changed my HAAS project twice completely and I'm fine with it as it is. I can (as it now is) keep it in my custom_components. So take the pull or leave it.

Reply:
I'm getting a bit lost on what to change now, as I keep changing it and still it seems to be going in the wrong direction. So same info on what I needed:

Rain Bird API allows to switch a valve on for a given duration.  So I need a switch + duration. There is no way to know which of the 4 - 8 -12 - 24 groups is connected. So that's why i wanted it configurable as most people won't have everything connected. 

Secondly I need to minimize the calls towards the API on the WiFi LNK module. When a call happens it takes about 10 seconds (sometimes longer) before a different connection can make requests, like the Rain Bird App. So as long as HAAS is the only one polling, it won't be an issue to have every switch polling. But most users will still want to use the app of Rain Bird itself and that is an issue. That's why I needed an additional layer/entity to handle requests and state.

For my own setup in HAAS I just use the Controller Entity and use customizations, scripts, automation scripts and custom sensors to interact with the 2 services exposed by this component. The switches are just there for people who want to have simple on / off sprinklers with a set duration in config.

So yes, I can make some changes, but I need to know exactly what it is you want.